File: macos/Guardian/AI/FrameQuality.swift
import CoreGraphics
import Foundation

/// Heuristic: is this frame essentially blank/uniform — i.e. there is nothing for the AI to see?
///
/// A failed capture, a screen-recording-protected window, or a solid wallpaper can come back as a
/// near-uniform frame. In those cases we cannot verify the content, so the caller handles it as
/// "can't see" (alert, never a hard block). Mirrors the Android `FrameQuality`.
enum FrameQuality {

    static func isUnreadable(_ image: CGImage) -> Bool {
        isUnreadableLuminance(sampleLuminance(image))
    }

    /// Coarse 32×32 grid sample of luminance (0...255).
    static func sampleLuminance(_ image: CGImage) -> [Int] {
        let w = image.width, h = image.height
        guard w > 0, h > 0 else { return [] }

        // Render into a known RGBA8 buffer so pixel reads are well-defined.
        let cols = min(32, w), rows = min(32, h)
        let bytesPerPixel = 4
        let bytesPerRow = cols * bytesPerPixel
        var buffer = [UInt8](repeating: 0, count: rows * bytesPerRow)
        let space = CGColorSpaceCreateDeviceRGB()
        guard let ctx = CGContext(
            data: &buffer, width: cols, height: rows, bitsPerComponent: 8,
            bytesPerRow: bytesPerRow, space: space,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        ) else { return [] }
        ctx.draw(image, in: CGRect(x: 0, y: 0, width: cols, height: rows))

        var lums = [Int]()
        lums.reserveCapacity(rows * cols)
        var i = 0
        while i < buffer.count {
            let r = Int(buffer[i]), g = Int(buffer[i + 1]), b = Int(buffer[i + 2])
            lums.append((r * 299 + g * 587 + b * 114) / 1000)
            i += bytesPerPixel
        }
        return lums
    }

    /// Pure heuristic over sampled luminance values (0...255). Extracted so it is unit-testable.
    static func isUnreadableLuminance(_ lums: [Int]) -> Bool {
        if lums.isEmpty { return true }
        var nearBlack = 0
        var sum: Int64 = 0
        var sumSq: Int64 = 0
        for lum in lums {
            if lum < 16 { nearBlack += 1 }
            sum += Int64(lum)
            sumSq += Int64(lum * lum)
        }
        let n = Double(lums.count)
        let mean = Double(sum) / n
        let variance = Double(sumSq) / n - mean * mean
        let blackFraction = Double(nearBlack) / n
        return blackFraction > 0.995 || variance < 4.0
    }
}

File: windows/guardian/ai/frame_quality.py
"""Port of `AI/FrameQuality.swift`.

Heuristic: is this frame essentially blank/uniform — i.e. there is nothing for the AI to see?

A failed capture, a capture-protected window (`SetWindowDisplayAffinity`/DRM), or a solid wallpaper
can come back as a near-uniform frame. In those cases we cannot verify the content, so the caller
handles it as "can't see" (alert, never a hard block). Thresholds are identical to the Mac/Android
versions.
"""


def sample_luminance(image) -> list:
    """Coarse 32x32 grid sample of luminance (0...255) from a PIL image."""
    w, h = image.size
    if w <= 0 or h <= 0:
        return []
    cols, rows = min(32, w), min(32, h)
    small = image.convert("RGB").resize((cols, rows))
    return [(r * 299 + g * 587 + b * 114) // 1000 for (r, g, b) in small.getdata()]


def is_unreadable_luminance(lums) -> bool:
    """Pure heuristic over sampled luminance values (0...255). Extracted so it is unit-testable."""
    if not lums:
        return True
    near_black = 0
    total = 0
    total_sq = 0
    for lum in lums:
        if lum < 16:
            near_black += 1
        total += lum
        total_sq += lum * lum
    n = float(len(lums))
    mean = total / n
    variance = total_sq / n - mean * mean
    black_fraction = near_black / n
    return black_fraction > 0.995 or variance < 4.0


def is_unreadable(image) -> bool:
    return is_unreadable_luminance(sample_luminance(image))
